---
title: "Week 7 report Group 8"
format: html
editor: visual
editor_options: 
  chunk_output_type: inline
---

# PCA tidyverse style on Gravier data

# Tasks

| Task                | Responsible | Done? |
|---------------------|-------------|-------|
| Create this file    | Dana        | Yes   |
| Load libraries      | Vakare      | Yes   |
| Load data           | Dana        | Yes   |
| Clean data          | Greta       | Yes   |
| Produce first plot  | Dana        | Yes   |
| Produce middle plot | Vakare      | Yes   |
| Reproduce last plot | Greta       | Yes   |

# Merging comment

Use system("git merge main") in R console to merge your branch to main. Remember to Pull before the merge.

# Load libraries

```{r}
#| warning: false
rm(list = ls())
library("tidyverse")
library("here")
library("gitcreds")
library("usethis")
library("broom")
library("cowplot")
```

# Set up git merging conflict code

```{r}
usethis::use_git_config(pull.rebase = "false")
```

# Download data the first time and save it ONLY RUN THIS ONCE IF YOU DO NOT HAVE THE DATA

```{r}
# target_url <- "https://github.com/ramhiser/datamicroarray/raw/master/data/gravier.RData"
# output_file <- "data/_raw/gravier.RData"
# curl::curl_download(url = target_url, destfile = output_file)
```

# Load data

```{r}
load(file = here("data/_raw/gravier.RData")) # loads as gravier
```

# Inspect data

```{r}
gravier |> names()
```

```{r}
gravier |> pluck("x") |> head(n = c(10,10))
```

```{r}
gravier |> pluck("y") |> head()
```

# Clean data

This is tidy data since each variable forms a column, each observation forms a row, and each value has a single cell.

```{r}
gravier_clean <- gravier |>
  bind_cols() |>
  as_tibble()
gravier_clean |> head(n = c(10,10))
```

```{r}
gravier_clean  |> 
  write_tsv(file = "data/02_gravier_clean.tsv.gz")
```

# First plot

**Running PCA (we need this for all 3 plots):**

```{r}
pca_fit <- gravier_clean |>
select(where(is.numeric)) |>      # retain only numeric columns as prcomp() can only work on numeric data
  prcomp(scale = TRUE)            # do PCA on scaled data (which has zero mean and unit variance for every variable i.e. every variable contributes equally to the PCA)
```

**Plotting the data in PC coordinates**

```{r}
pca_fit |>
  augment(gravier_clean) |> # add original dataset back in
  ggplot(aes(.fittedPC1, .fittedPC2, color = y)) + 
  geom_point(size = 1.5) +
    labs(title = "Exploratory PC coordinates plot", 
         x = "Fitted PC1", y = "Fitted PC2", 
         color = "Cancer outcome") +
  scale_color_manual(values = c(good = "hotpink", poor = "brown"))  +
  theme_half_open(12) + background_grid()
```

# Second plot

Investigating how 10 random genes contribute to first two principal components (i.e. is a gene increasing or decreasing PC value?).

```{r}
pca_long <- pca_fit |>
  tidy(matrix = "rotation")
pca_long |>
  head(n =10)
```

```{r}
# Turning long data into wide data
pca_wide <- pca_long |>
  pivot_wider(names_from = PC, values_from = value)

# picking 10 random genes to plot
set.seed(5606)
pca_sample <- pca_wide |>
  slice_sample(n = min(10, nrow(pca_wide)))

# Defining arrow style
arrow_style <- arrow(angle = 20, ends = "first", type = "closed", 
                     length = grid::unit(5, "pt"))
```

```{r}
# Plotting selected data
ggplot(pca_sample, aes(x = `1`, y = `2`)) + # 1 is PC1, 2 is PC2
  geom_segment(aes(xend = 0, yend = 0), 
               arrow = arrow_style, alpha = 0.7, color = "hotpink") +
  geom_text(aes(label = column),
            hjust = 1,
            nudge_x = 0.007,
            color = "purple",
            size = 3,
            fontface = "bold") +
  coord_fixed() +
  theme_half_open(12) + 
  background_grid() +
  labs(title = "PC coordinates rotation matrix",
       x = "PC1",
       y = "PC2")
```

# Third plot

## Variance explained by each PC (up to PC10)

In this plot we are looking at the variance explained by each principle component (up to PC10).

```{r}
# get tibble of PCs and variances
pca_fit |>
  tidy(matrix = "eigenvalues") |>
# choose only the first 10 PCs 
  filter(PC <= 10) |>
  ggplot(aes(x = PC, y = percent)) + 
  geom_col(fill= "pink", alpha = 0.7) +
  scale_x_continuous(breaks = 1:10) +
  scale_y_continuous(labels = scales::percent,
                     expand = expansion(mult = c(0, 0.05))) +
  theme(plot.title = element_text(size = 10)) + 
  labs(title = "Variation of the First 10 Principal Components",
    x = "Principal components", 
    y = 'Percent') +
  theme_half_open(12) + 
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.y = element_line(color = "gray50", linewidth = 0.2))
```

The first component captures more than 8 percent of the variation in the data. The remaining nine components, among the first ten, capture progressively smaller percentages of the variation. Although PC1 capturing 8% is a low number, this is typical for large gene expression datasets and thus as expected.
